Tests/SwiftGRPCNIOTests/NIOServerTests.swift

```diff
@@ -0,0 +1,302 @@
+    }
+    return sendOperationChain.map { GRPCStatus.ok }
+  }
+
+  func update(context: StreamingResponseCallContext<Echo_EchoResponse>) -> EventLoopFuture<(StreamEvent<Echo_EchoRequest>) -> Void> {
+    var sendOperationChain = context.eventLoop.newSucceededFuture(result: ())
+    var count = 0
+    return context.eventLoop.newSucceededFuture(result: { event in
+      switch event {
+      case .message(let message):
+        var response = Echo_EchoResponse()
+        response.text = "Swift echo update (\(count)): \(message.text)"
+        sendOperationChain = sendOperationChain.then { context.sendResponse(response) }
+        count += 1
+        
+      case .end:
+        sendOperationChain
+          .map { GRPCStatus.ok }
+          .cascade(promise: context.statusPromise)
+      }
+    })
+  }
+}
+
+class NIOServerTests: NIOServerTestCase {
+  static var allTests: [(String, (NIOServerTests) -> () throws -> Void)] {
+    return [
+      ("testUnary", testUnary),
+      ("testUnaryLotsOfRequests", testUnaryLotsOfRequests),
+      ("testClientStreaming", testClientStreaming),
+      ("testClientStreamingLotsOfMessages", testClientStreamingLotsOfMessages),
+      ("testServerStreaming", testServerStreaming),
+      ("testServerStreamingLotsOfMessages", testServerStreamingLotsOfMessages),
+      ("testBidirectionalStreamingBatched", testBidirectionalStreamingBatched),
+      ("testBidirectionalStreamingPingPong", testBidirectionalStreamingPingPong),
+      ("testBidirectionalStreamingLotsOfMessagesBatched", testBidirectionalStreamingLotsOfMessagesBatched),
+      ("testBidirectionalStreamingLotsOfMessagesPingPong", testBidirectionalStreamingLotsOfMessagesPingPong)
+    ]
+  }
+
+  static let lotsOfStrings = (0..<1000).map { String(describing: $0) }
+
+  var eventLoopGroup: MultiThreadedEventLoopGroup!
+  var server: GRPCServer!
+
+  override func setUp() {
+    super.setUp()
+
+    // This is how a GRPC server would actually be set up.
+    eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+    server = try! GRPCServer.start(
+      hostname: "localhost", port: 5050, eventLoopGroup: eventLoopGroup, serviceProviders: [EchoProvider_NIO()])
+      .wait()
+  }
+
+  override func tearDown() {
+    try! server.close().wait()
+
+    try! eventLoopGroup.syncShutdownGracefully()
+    eventLoopGroup = nil
+
+    super.tearDown()
+  }
+}
+
+extension NIOServerTests {
+  func testUnary() {
+    XCTAssertEqual("Swift echo get: foo", try! client.get(Echo_EchoRequest(text: "foo")).text)
+  }
+
+  func testUnaryLotsOfRequests() {
+    // Sending that many requests at once can sometimes trip things up, it seems.
+    client.timeout = 5.0
+    let clockStart = clock()
+    let numberOfRequests = 1_000  //! FIXME: If we set this higher, it causes a crash related to `StreamManager.maxCachedStreamIDs`.
+    for i in 0..<numberOfRequests {
+      if i % 1_000 == 0 && i > 0 {
```

Comment:
Why the `% 1_000` test here if `numberOfRequest == 1_000`?

Reply:
Good point. I have increased this to 2000 now, as the underlying error got fixed by Corey.